pkg/aws/commandbuilder.go

```diff
@@ -0,0 +1,126 @@
+package aws
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+	"strings"
+)
+
+type Service string
+
+const (
+	UNDEFINED_SERVICE Service = "undefined service"
+	IAM               Service = "iam"
+)
+
+type Command string
+
+const (
+	UNDEFINED_COMMAND                Command = "undefined command"
+	CREATE_ROLE                      Command = "create-role"
+	DELETE_ROLE                      Command = "delete-role"
+	CREATE_POLICY                    Command = "create-policy"
+	DELETE_POLICY                    Command = "delete-policy"
+	CREATE_POLICY_VERSION            Command = "create-policy-version"
+	DELETE_ROLE_POLICY               Command = "delete-role-policy"
+	ATTACH_ROLE_POLICY               Command = "attach-role-policy"
+	DETACH_ROLE_POLICY               Command = "detach-role-policy"
+	TAG_POLICY                       Command = "tag-policy"
+	TAG_ROLE                         Command = "tag-role"
+	CREATE_OPEN_ID_CONNECT_PROVIDER  Command = "create-open-id-connect-provider"
+	DELETE_OPEN_ID_CONNECT_PROVIDER  Command = "delete-open-id-connect-provider"
+	DELETE_ROLE_PERMISSIONS_BOUNDARY Command = "delete-role-permissions-boundary"
+)
+
+type Param string
+
+const (
+	UNDEFINED_PARAM              Param = "undefined param"
+	TAGS                         Param = "tags"
+	ROLE_NAME                    Param = "role-name"
+	ASSUME_ROLE_POLICY_DOCUMENT  Param = "assume-role-policy-document"
+	PERMISSIONS_BOUNDARY         Param = "permissions-boundary"
+	PATH                         Param = "path"
+	POLICY_NAME                  Param = "policy-name"
+	POLICY_DOCUMENT              Param = "policy-document"
+	POLICY_ARN                   Param = "policy-arn"
+	URL                          Param = "url"
+	CLIENT_ID_LIST               Param = "client-id-list"
+	THUMBPRINT_LIST              Param = "thumbprint-list"
+	OPEN_ID_CONNECT_PROVIDER_ARN Param = "open-id-connect-provider-arn"
+	SET_AS_DEFAULT               Param = "set-as-default"
+)
+
+type ICommandBuilder interface {
+	SetService(awsService Service)
+	SetCommand(awsCommand Command)
+	AddParam(awsParam Param, value interface{})
+	ResetParams()
+	Build() string
+}
+
+type CommandBuilder struct {
+	service string
+	command string
+	params  []string
+}
+
+type StringMap map[string]string
+
+func (b *CommandBuilder) SetService(awsService Service) {
+	b.service = fmt.Sprint(awsService)
+}
+
+func (b *CommandBuilder) SetCommand(awsCommand Command) {
+	b.command = fmt.Sprint(awsCommand)
+}
+
+func (b *CommandBuilder) AddParam(awsParam Param, value interface{}) {
+	if b.params == nil {
+		b.params = make([]string, 1)
+	}
+	stringValue := ""
+	typeOf := fmt.Sprintf("%T", value)
+	switch typeOf {
+	case "string":
+		stringValue = value.(string)
+	case "map[string]string":
+		stringValue = createKeyValuePairs(value.(map[string]string))
+	default:
+		stringValue = ""
```

Comment:
We could just error out too as the default, type not supported or something.
The point is to make sure someone would need to implement it as we don't have this specific case yet

Reply:
Ended up getting the string value with %v, this will cover most of the base cases.
Is there a guideline to have some form of try catch like, didn't want to error as would need to check everytime we add a param